tensorflow/contrib/layers/kernels/layer_norm_fused_op.cc

```diff
@@ -0,0 +1,317 @@
+struct LaunchLayerNormBiasAddOp;
+
+template <typename Device, typename T>
+struct LaunchLayerNormFusedOp;
+
+#if GOOGLE_CUDA
+template <typename T>
+struct LayerNormGPULaunch {
+  static void Run(const GPUDevice& d, const LayerNormFusedArgs args,
+                  const T* input, T* output);
+};
+template <typename T>
+struct LaunchLayerNormOp<GPUDevice, T> {
+  static void launch(OpKernelContext* ctx, const LayerNormFusedArgs args,
+                     const T* input, T* output) {
+    const GPUDevice& d = ctx->eigen_device<GPUDevice>();
+    LayerNormGPULaunch<T>().Run(d, args, input, output);
+  }
+};
+
+template <typename T>
+struct LayerNormBiasAddGPULaunch {
+  static void Run(const GPUDevice& d, const LayerNormFusedArgs args,
+                  const T* input, const T* beta, T* output);
+};
+template <typename T>
+struct LaunchLayerNormBiasAddOp<GPUDevice, T> {
+  static void launch(OpKernelContext* ctx, const LayerNormFusedArgs args,
+                     const T* input, const T* beta, T* output) {
+    const GPUDevice& d = ctx->eigen_device<GPUDevice>();
+    LayerNormBiasAddGPULaunch<T>().Run(d, args, input, beta, output);
+  }
+};
+
+template <typename T>
+struct LayerNormFusedGPULaunch {
+  static void Run(const GPUDevice& d, const LayerNormFusedArgs args,
+                  const T* input, const T* gamma, const T* beta, T* output);
+};
+template <typename T>
+struct LaunchLayerNormFusedOp<GPUDevice, T> {
+  static void launch(OpKernelContext* ctx, const LayerNormFusedArgs args,
+                     const T* input, const T* gamma, const T* beta, T* output) {
+    const GPUDevice& d = ctx->eigen_device<GPUDevice>();
+    LayerNormFusedGPULaunch<T>().Run(d, args, input, gamma, beta, output);
+  }
+};
+#endif  // GOOGLE_CUDA
+
+template <typename Device, typename T>
+class LayerNormOp : public OpKernel {
+ public:
+  explicit LayerNormOp(OpKernelConstruction* context) : OpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("epsilon", &epsilon_));
+  }
+
+  void Compute(OpKernelContext* context) override {
+    const Tensor& input = context->input(0);
+
+    OP_REQUIRES(
+        context, input.dims() >= 2,
+        errors::InvalidArgument("input dimensions must be larger than 2D",
+                                input.shape().DebugString()));
+
+    const int32 last_dim = input.dims() - 1;
+    const int32 depth = input.dim_size(last_dim);
+
+    int32 n_slices = 1;
+    for (int i = 0; i < last_dim; ++i) {
+      n_slices *= input.dim_size(i);
+    }
+
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK(context,
+                   context->allocate_output(0, input.shape(), &output));
+
+    VLOG(2) << "LayerNormCustom: "
+            << "depth:" << depth << ", "
+            << "n_slices:" << n_slices;
+
+    LayerNormFusedArgs args;
+    args.depth = depth;
+    args.n_slices = n_slices;
+    args.n_inputs = n_slices * depth;
+    args.epsilon = epsilon_;
+
+    if (depth <= WARP_SIZE) {
```

Comment:
Sorry, I am not sure what Cuda translation unit means. Do you mean I should move this code to `layer_norm_fused_op_gpu.cu.cc`?